octodns/provider/base.py

```diff
@@ -94,9 +95,56 @@ def _process_desired_zone(self, desired):
                 record = record.copy()
                 record.values = [record.value]
                 desired.add_record(record, replace=True)
+            elif record._type == 'NS' and record.name == '':
+                if self.SUPPORTS_ROOT_NS:
+                    # record that we saw a root NS record
+                    have_root_ns = True
+                else:
+                    # ignore, we can't manage root NS records
+                    msg = \
+                        f'root NS record not supported for {record.fqdn}'
+                    fallback = 'ignoring it'
+                    self.supports_warn_or_except(msg, fallback)
+                    desired.remove_record(record)
+
+        if self.SUPPORTS_ROOT_NS and not have_root_ns:
+            raise SupportsException(f'{self.id}: provider supports root NS '
+                                    'record management, but no record '
+                                    f'configured in {desired.name}; aborting')
```

Comment:
So this is the most noteworthy change here. It seems like in order to safely start supporting root NS management we need to require that one is sourced/configured if any of the targets `SUPPORTS_ROOT_NS = True`, otherwise a plan could be generated that removes the root NS records.

This would potentially be a ~breaking change as in if anyone updates to an octoDNS and provider that `SUPPORTS_ROOT_NS = True` and isn't including a root NS record in their zone's desired state they'd get an exception. It's not breaking as in it would cause bad state or anything like that (it's trying to prevent that.)

We could leave out this check and instead rely on people noticing that the root NS is getting deleted, but that seems worrying. 

Thoughts anyone?

Reply:
Let's keep the check and prevent people's NS records blowing up.